### ISSUE TITLE
Add Lightning-like Profiler to Keras

### ISSUE THREAD

**Issue by innat** [keras-team/keras#19023]
Reopening from [here](https://github.com/keras-team/tf-keras/issues/350). 

## Describe the feature and the current behavior/state

In PyTorch-Lightning, you can identify bottlenecks in your code using [**Profiler**](https://pytorch-lightning.readthedocs.io/en/1.6.0/advanced/profiler.html) API. And I've found it super useful and easy to inspect the pros and cons of the code. This API offers 3 type of inferface, namely `simple`, `advance` and `pytorch`. I mostly use `simple` that summarize the execution time of medium to high level training + validation + inference API. The saved log files look something as follows. 

```python
# by passing a string
trainer = Trainer(..., profiler="simple")

# or by passing an instance
from pytorch_lightning.profiler import SimpleProfiler

profiler = SimpleProfiler()
trainer = Trainer(..., profiler=profiler)
```

![a](https://user-images.githubusercontent.com/17668390/204000492-0e7c8ee3-31ba-4194-bde1-acca200b2a23.png)

I don't know, in `keras`, if we have anything like this already. By having such high-level functionality in `keras` would really great. It will help to inspect the bottleneck of dataloader, model, callbacks etc, by reporting the time and memory consumption in the system.

I looked for it before, [discussion](https://discuss.tensorflow.org/t/profiler-report-like-pytorch-lighting-in-tf-keras-interface/3768). 
 
## Will this change the current api? How?

This can be considered as a [Callback](https://keras.io/api/callbacks/), I think. 

```python
keras.callbacks.Profiler
or
keras.utils.Profiler
```

**Who will benefit from this feature?**

The keras community. Engineers and researchers, both. Mostly engineers in production phase for code optimization.

**[Contributing](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md)**

- Do you want to contribute a PR? (yes/no):
- If yes, please read [this page](https://github.com/keras-team/keras/blob/master/CONTRIBUTING.md) for instructions
- Briefly describe your candidate solution(if contributing):

**Comment by nkovela1** [comment]
Hi @innat, this would be a good addition as a Callback! Would you be willing to contribute a PR to add this Profiler Callback?

**Comment by innat** [comment]
Yes, I'm willing to contribute. Before that, could you please give some thought on [this](https://github.com/keras-team/tf-keras/issues/350#issuecomment-1731852799). Thanks.

**Comment by fchollet** [comment]
> Before that, could you please give some thought on https://github.com/keras-team/tf-keras/issues/350#issuecomment-1731852799. Thanks.

There are no profiling plans from our side at this time -- if you want to create this feature, you're welcome to do it, it would be a very valuable addition.

**Comment by ghsanti** [comment]
I've [tried something here](https://github.com/tensorflow/profiler/issues/602#issuecomment-2264427977)I don't know if it's a meaningful plot yet, basically I'm using **Trace** in a loop:


**CODE**

```py
# for colab
# !pip uninstall tensorflow keras tensorboard-plugin-profile tensorboard tb-nightly tensorboardX -y
# !pip install -U tensorflow==2.16.1 keras==3.4.1 tensorboard-plugin-profile

import tensorflow as tf

from datetime import datetime
import keras
from keras.api.layers import Dense, Flatten
from keras.api.callbacks import TensorBoard

import tensorflow_datasets as tfds

tfds.disable_progress_bar()


(ds_train, ds_test), ds_info = tfds.load(
    "mnist",
    split=["train", "test"],
    shuffle_files=True,
    as_supervised=True,
    with_info=True,
)


def normalize_img(image, label):
    """Normalizes images: `uint8` -> `float32`."""
    return tf.cast(image, tf.float32) / 255.0, label


ds_train = ds_train.map(normalize_img)
ds_train = ds_train.batch(128)
ds_train = ds_train.cache()
ds_train = ds_train.prefetch(tf.data.experimental.AUTOTUNE)

ds_test = ds_test.map(normalize_img)
ds_test = ds_test.batch(128)
ds_test = ds_test.cache()
ds_test = ds_test.prefetch(tf.data.experimental.AUTOTUNE)


model = keras.models.Sequential(
    [
        keras.layers.Input(shape=(28, 28, 1)),
        keras.layers.Flatten(),
        keras.layers.Dense(412, activation="relu"),
        keras.layers.Dense(10, activation="softmax"),
    ]
)
model.compile(
    loss="sparse_categorical_crossentropy",
    optimizer=keras.optimizers.Adam(0.001),
    metrics=["accuracy"],
)

logs = "logs/" + datetime.now().strftime("%Y%m%d-%H%M%S")

tboard_callback = keras.callbacks.TensorBoard(
    log_dir=logs,
    histogram_freq=1,
)

tf.profiler.experimental.start(logs)
for step in range(6):
    # Creates a trace event for each training step with the step number.
    with tf.profiler.experimental.Trace("Train", step_num=step, _r=1):
        model.fit(
            ds_train,
            batch_size=16,
            callbacks=[tboard_callback],
            validation_data=ds_test,
        )
tf.profiler.experimental.stop()
```
